### DISCUSSION TITLE
How to install the C++ binary distribution in Visual Studio 2022 (Win10)?

### DISCUSSION THREAD

**Discussion by zeldaxlove64** [google/or-tools#3963] [Build and configuration issues] [answered]
Windows 10 21H2 19044.1645
Microsoft Visual Studio Enterprise 2022 (64-bit) - Current Version 17.7.6
or-tools_x64_VisualStudio2022_cpp_v9.7.2996

I have been using the python version of or-tools before, and now I want to try the c++ version. But I tried for two days and search a lot of information on the Internet, but I still couldn't get the program to run. I don’t know which one I did wrong or what steps I missed. I hope you can help me figure it out.

Details are as follows:
I download the binary distribution: [or-tools_x64_VisualStudio2022_cpp_v9.7.2996.zip](https://github.com/google/or-tools/releases/download/v9.7/or-tools_x64_VisualStudio2022_cpp_v9.7.2996.zip) and extracted it to the directory: "E:\Microsoft Visual Studio\Libs\or-tools_x64_VisualStudio2022_cpp_v9.7.2996".

I create a new `Empty Project (C++)` and name it "TestProject" in Visual Studio, and the project will be created in "E:\Microsoft Visual Studio\Project\TestProject\".

I `Add > New Item` and name it "program.cc" as file name in the project, and I copy and paste the code which is in [Get Started with OR-Tools for C++](https://developers.google.com/optimization/introduction/cpp) into the "program.cc" file.

I right-clicked the project to open `Properties` and modified the following settings for both `Configuration: Release` and `Configuration: Debug`:
Set `Configuration Properties > General > C++ Language Standard` to `ISO C++17 Standard (/std:c++17)`.
Set `Configuration Properties > C/C++ > Code Generation > Runtime Library` to `Multi-threaded DLL (/MD)`.
Add `E:\Microsoft Visual Studio\Libs\or-tools_x64_VisualStudio2022_cpp_v9.7.2996\include` into `Configuration Properties > C/C++ > General > Additional Include Directories`.
Add `E:\Microsoft Visual Studio\Libs\or-tools_x64_VisualStudio2022_cpp_v9.7.2996\lib` into `Configuration Properties > Linker > General > Additional Library Directories`.
Add `ortools_full.lib` into `Configuration Properties > Linker > Input > Additional Dependencies`.

I select `Solution Configurations` as `Release`, and I select `Solution Platforms` as `x64`, and I press `▶ Local Windows Debugger` so the program starts running.

And I get 7 Errors, 200 Warnings, 0Messages.
**Error List:**
**Severity | Code | Description | Project | File | Line | Suppression State**
Error	LNK2001	unresolved external symbol "bool __cdecl utf8_range::IsStructurallyValid(class std::basic_string_view<char,struct std::char_traits<char> >)" (?IsStructurallyValid@utf8_range@@YA_NV?$basic_string_view@DU?$char_traits@D@std@@@std@@@Z)	TestProject	E:\Microsoft Visual Studio\Project\TestProject\ortools_full.lib(parse_context.obj)	1	
Error	LNK2001	unresolved external symbol "bool __cdecl utf8_range::IsStructurallyValid(class std::basic_string_view<char,struct std::char_traits<char> >)" (?IsStructurallyValid@utf8_range@@YA_NV?$basic_string_view@DU?$char_traits@D@std@@@std@@@Z)	TestProject	E:\Microsoft Visual Studio\Project\TestProject\ortools_full.lib(generated_message_tctable_lite.obj)	1	
Error	LNK2001	unresolved external symbol "bool __cdecl utf8_range::IsStructurallyValid(class std::basic_string_view<char,struct std::char_traits<char> >)" (?IsStructurallyValid@utf8_range@@YA_NV?$basic_string_view@DU?$char_traits@D@std@@@std@@@Z)	TestProject	E:\Microsoft Visual Studio\Project\TestProject\ortools_full.lib(wire_format_lite.obj)	1	
Error	LNK2001	unresolved external symbol "bool __cdecl utf8_range::IsStructurallyValid(class std::basic_string_view<char,struct std::char_traits<char> >)" (?IsStructurallyValid@utf8_range@@YA_NV?$basic_string_view@DU?$char_traits@D@std@@@std@@@Z)	TestProject	E:\Microsoft Visual Studio\Project\TestProject\ortools_full.lib(untyped_message.obj)	1	
Error (active)	E1097	unknown attribute "no_sanitize_address"	TestProject	E:\Microsoft Visual Studio\Libs\or-tools_x64_VisualStudio2022_cpp_v9.7.2996\include\absl\strings\internal\cord_internal.h	740	
Error (active)	E0317	return type is not identical to nor covariant with return type "google::protobuf::internal::UntypedMapBase *" of overridden virtual function "google::protobuf::internal::MapFieldBaseForParse::MutableMap"	TestProject	E:\Microsoft Visual Studio\Libs\or-tools_x64_VisualStudio2022_cpp_v9.7.2996\include\google\protobuf\map_field.h	626	
Error	LNK1120	1 unresolved externals	TestProject	E:\Microsoft Visual Studio\Project\TestProject\x64\Release\TestProject.exe	1	

**Output:**
Build started...
1>------ Build started: Project: TestProject, Configuration: Release x64 ------
1>ortools_full.lib(parse_context.obj) : error LNK2001: unresolved external symbol "bool __cdecl utf8_range::IsStructurallyValid(class std::basic_string_view<char,struct std::char_traits<char> >)" (?IsStructurallyValid@utf8_range@@YA_NV?$basic_string_view@DU?$char_traits@D@std@@@std@@@Z)
1>ortools_full.lib(generated_message_tctable_lite.obj) : error LNK2001: unresolved external symbol "bool __cdecl utf8_range::IsStructurallyValid(class std::basic_string_view<char,struct std::char_traits<char> >)" (?IsStructurallyValid@utf8_range@@YA_NV?$basic_string_view@DU?$char_traits@D@std@@@std@@@Z)
1>ortools_full.lib(wire_format_lite.obj) : error LNK2001: unresolved external symbol "bool __cdecl utf8_range::IsStructurallyValid(class std::basic_string_view<char,struct std::char_traits<char> >)" (?IsStructurallyValid@utf8_range@@YA_NV?$basic_string_view@DU?$char_traits@D@std@@@std@@@Z)
1>ortools_full.lib(untyped_message.obj) : error LNK2001: unresolved external symbol "bool __cdecl utf8_range::IsStructurallyValid(class std::basic_string_view<char,struct std::char_traits<char> >)" (?IsStructurallyValid@utf8_range@@YA_NV?$basic_string_view@DU?$char_traits@D@std@@@std@@@Z)
1>E:\Microsoft Visual Studio\Project\TestProject\x64\Release\TestProject.exe : fatal error LNK1120: 1 unresolved externals
1>Done building project "TestProject.vcxproj" -- FAILED.
========== Build: 0 succeeded, 1 failed, 0 up-to-date, 0 skipped ==========
========== Build started at 14:33 and took 00.914 seconds ==========

Please help me, thank you.

**Comment by lperron** [comment, marked as answer]
crossposted from https://stackoverflow.com/questions/77365562/how-to-install-the-or-tools-c-binary-distribution-in-visual-studio-2022-win10
